### PR TITLE
Incremental training

### PR DESCRIPTION
**Pull request by nsorros**
This PR reintroduces the idea of incremental learning which we first wanted to use to train our Tfidf-SVM baseline.

This time we use our CNN model which leverages `tf.data.Dataset` to implement incremental learning. Note that tensorflow also allows for a generator to be passed in but in that case functionally such as batching, shuffling etc would need to be implemented whereas now it comes for free.

Note that `KerasVectorizer` and `MultiLabelBinarizer` allow for a generator to be passed in so training them efficiently is relatively easy. One thing we need to work around, as is often in this project, is sparse labels. In particular the dataset converts to dense as it returns data but this means that test_data need to be converted back to sparse to fit into memory.

### CODE REVIEW

**Comment by nsorros** [comment]
[Train CNN on entire MeSH](https://trello.com/c/GTAM7HL6/1174-train-cnn-on-entire-mesh)

**Comment by nsorros** [comment]
Just a reminder about this PR @lizgzil @aCampello

File: requirements.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
 scikit-multilearn
 arff
 python-dateutil<2.8.1,>=2.1
--e git+git://github.com/wellcometrust/WellcomeML.git@ca5d1c38f28a50aacdefa5b5b47647389194e826#egg=wellcomeml[deep-learning]
+-e git+git://github.com/wellcometrust/WellcomeML.git@14668e8650516363453a85bc2a5546cfe8811951#egg=wellcomeml[deep-learning]
```
Comment (aCampello) [inline requirements.txt:4]: Do you not want to use the pypi version?

**Review by aCampello — APPROVED** [review]
Looks good. Wondering why you are calling 'incremental learning' is it because you are loading the data at every batch ? I thought about something like: when you get more annotated data, you can pick the model from where you stopped and continue training..

File: grants_tagger/utils.py [hunk 2]
```diff
@@ -95,3 +68,78 @@ def calc_performance_per_tag(Y_true, Y_pred, tags):
...
+        text_encoded = tokenizer.transform(texts)
+        tags_encoded = label_binarizer.transform(tags)
+
+        if sparse_labels:
+            tags_encoded = tags_encoded.todense() # returns matrix
+            tags_encoded = np.asarray(tags_encoded)
+        
+        return text_encoded, tags_encoded
+
+    def data_gen():
```
Comment (lizgzil) [inline grants_tagger/utils.py:96]: it would be nice for some comments on this or a function rename, since it isnt just a data generator but a kind of windowed data generator?

File: requirements.txt (hunk 1 above)
Comment (nsorros) [reply requirements.txt:4]: haha, because I so often submit a feature and fix I keep pointing to the latest commit to develop new features and use the latest published when I "publish" a new version

File: grants_tagger/utils.py (hunk 2 above)
Comment (nsorros) [reply grants_tagger/utils.py:96]: I can add comments for sure, guilty on that, need to write more docstrings 😬 

this is actually a callable generator that returns one example at a time. why callable? because when it returns all elements tensorflow needs a way to restart so it basically calls it again, otherwise it would be able to iterate only once.

```
def data_gen():
    for i in range(5):
        yield I

non_callable_generator = data_gen()
callable_generator = data_gen
```

**Comment by nsorros** [comment]
> Looks good. Wondering why you are calling 'incremental learning' is it because you are loading the data at every batch ? I thought about something like: when you get more annotated data, you can pick the model from where you stopped and continue training..

I was looking for a good name and settled with incremental learning, the older name was online_learning. I am trying to describe situation where you can update an algorithm with some examples which not algorithms allow, in this case you can leverage "incremental learning" to not load all data into memory and give them one by one. Possibly another name could be related to the memory efficiency thing.

see also: https://scikit-learn.org/0.15/modules/scaling_strategies.html
